### PR TITLE
Fixes hyperdash computation (for nested objects)

### LINKED ISSUE #2586: Hyper dash generation is incorrect on JuiceStreams
When I'm trying to do implements on catch difficulty calculation, I noticed that there's a serious issue on hyper calculating, since in lazer codes droplets are changed to **JuiceStreams**, the hypers' are not being correctly calculated in **CatchBeatmapProcessor** since the processor is from old codes.

Listing the issue here and I'm trying to fix it.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/58565253-hyper-dash-generation-is-incorrect-on-juicestreams?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Hanamuke**
Fix #2586 
Rewrite from #2610 
Fixes hyperdash computation to take juice streams into account correctly.
I tested the patch, with high/low CS, converted/native and did not notice anything fondamently wrong.

### CODE REVIEW

File: osu.Game.Rulesets.Catch/Beatmaps/CatchBeatmapProcessor.cs [hunk 1]
```diff
@@ -74,42 +74,43 @@ private void applyPositionOffsets()
...
+                if (currentObject is Fruit)
+                    objectWithDroplets.Add(currentObject);
+                if (currentObject is JuiceStream)
+                    foreach (var currentJuiceElement in currentObject.NestedHitObjects)
+                        if (!(currentJuiceElement is TinyDroplet))
+                            objectWithDroplets.Add((CatchHitObject)currentJuiceElement);
+            }
 
-                CatchHitObject nextObject = objects[i + 1];
+            double halfCatcherWidth = CatcherArea.CATCHER_SIZE * objectWithDroplets[0].Scale / CatchPlayfield.BASE_WIDTH / 2;
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Catch/Beatmaps/CatchBeatmapProcessor.cs]: This is being done in multiple locations now. Expose a static method from `CatcherArea`, `GetCatcherSize(BeatmapDifficulty difficulty)` and pass `Beatmap.BeatmapInfo.BaseDifficulty` available here in.

Moving forward, we need to keep catcher-specific calculations within `CatcherArea`/`Catcher`.

File: osu.Game.Rulesets.Catch/Objects/CatchHitObject.cs [hunk 2]
```diff
@@ -27,7 +27,9 @@ public abstract class CatchHitObject : HitObject, IHasXPosition, IHasComboInform
         public int ComboIndex { get; set; }
 
         /// <summary>
-        /// The distance for a fruit to to next hyper if it's not a hyper.
+        /// The difference between the distance of the next object
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Catch/Objects/CatchHitObject.cs]: "Difference between the distance to the next object and the distance that would have triggered a hyper dash. A value close to 0 indicates a difficult jump (for difficulty calculation)."

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Logic looks good in general, just a few minor fixups.

**Comment by Hanamuke** [comment]
I was concerned about Fruit that were during a JuiceStream, as it happens in some freestyle map, so i sorted the objects.

**Review by smoogipoo — APPROVED** [review]
Looks good to me for an initial implementation 👍

**Pull request merged**